src/network/operator.rs

```diff
@@ -0,0 +1,335 @@
+/// Shorthand for the child half of the `Operator` message channel.
+type OperatorHandler<N> = mpsc::Receiver<OperatorRequest<N>>;
+
+///
+/// An enum of requests that the `Operator` struct processes.
+///
+#[derive(Debug)]
+pub enum OperatorRequest<N: Network> {
+    /// PoolRegister := (peer_ip, worker_address)
+    PoolRegister(SocketAddr, Address<N>),
+    /// PoolResponse := (peer_ip, proposed_block, worker_address)
+    PoolResponse(SocketAddr, Block<N>, Address<N>),
+}
+
+/// The predefined base share difficulty.
+const BASE_SHARE_DIFFICULTY: u64 = u64::MAX / 2u64.pow(1);
+
+///
+/// An operator for a program on a specific network in the node server.
+///
+#[derive(Debug)]
+pub struct Operator<N: Network, E: Environment> {
+    /// The address of the operator.
+    address: Option<Address<N>>,
+    /// The local address of this node.
+    local_ip: SocketAddr,
+    /// The state storage of the operator.
+    state: Arc<OperatorState<N>>,
+    /// The current block template that is being mined on by the operator.
+    block_template: RwLock<Option<BlockTemplate<N>>>,
+    /// A list of provers and their associated state := (last_submitted, share_difficulty, shares_submitted_since_reset)
+    provers: RwLock<HashMap<Address<N>, (Instant, u64, u32)>>,
+    /// The operator router of the node.
+    operator_router: OperatorRouter<N>,
+    /// The pool of unconfirmed transactions.
+    memory_pool: Arc<RwLock<MemoryPool<N>>>,
+    /// The peers router of the node.
+    peers_router: PeersRouter<N, E>,
+    /// The ledger state of the node.
+    ledger_reader: LedgerReader<N>,
+    /// The ledger router of the node.
+    ledger_router: LedgerRouter<N>,
+    /// The prover router of the node.
+    prover_router: ProverRouter<N>,
+}
+
+impl<N: Network, E: Environment> Operator<N, E> {
+    /// Initializes a new instance of the operator.
+    pub async fn open<S: Storage, P: AsRef<Path> + Copy>(
+        tasks: &Tasks<JoinHandle<()>>,
+        path: P,
+        address: Option<Address<N>>,
+        local_ip: SocketAddr,
+        memory_pool: Arc<RwLock<MemoryPool<N>>>,
+        peers_router: PeersRouter<N, E>,
+        ledger_reader: LedgerReader<N>,
+        ledger_router: LedgerRouter<N>,
+        prover_router: ProverRouter<N>,
+    ) -> Result<Arc<Self>> {
+        // Initialize an mpsc channel for sending requests to the `Operator` struct.
+        let (operator_router, mut operator_handler) = mpsc::channel(1024);
+
+        // Initialize the operator.
+        let operator = Arc::new(Self {
+            address,
+            local_ip,
+            state: Arc::new(OperatorState::open_writer::<S, P>(path)?),
+            block_template: RwLock::new(None),
+            provers: RwLock::new(HashMap::new()),
+            operator_router,
+            memory_pool,
+            peers_router,
+            ledger_reader,
+            ledger_router,
+            prover_router,
+        });
+
+        if E::NODE_TYPE == NodeType::Operator {
```

Comment:
(this also applies to line 144): is it possible to call `Operator::open` without the node being an operator? wondering if this predicate is needed

Reply:
Yes it is possible